### ISSUE TITLE
SVT: Failed to delete a Default WebSphere Liberty MicroProfile project after executing a load test, which produced profiling data

### ISSUE THREAD

**Issue by sujeilyfonseca** [eclipse/codewind#2456]
**Codewind version:** 0.10.0
**OS:** VM with Windows 10

**IDE extension version:** 0.10.0
**IDE version:** 1.42.0

Build date: 2020-03-10-12:22:59
build_info.url=https://download.eclipse.org/codewind/codewind-vscode/0.10.0/7
build_info.che.SHA-1=a1bfcb9274bc5365489e96596ab2faea4dc1678f
build_info.SHA-1=90663a8ff7ba33857f163016cb5c813df2d2e35d

**Description:** Failed to delete a Default WebSphere Liberty MicroProfile project after executing a load test, which produced profiling data (Codewind Java Profiler)

![Screen Shot 2020-03-11 at 10 46 25 AM](https://user-images.githubusercontent.com/49531960/76429540-92a04a80-6385-11ea-8921-8f255d2a477d.png)

![Screen Shot 2020-03-11 at 10 43 28 AM](https://user-images.githubusercontent.com/49531960/76429458-72708b80-6385-11ea-8a84-8da2f7287b31.png)

![Screen Shot 2020-03-11 at 10 45 17 AM](https://user-images.githubusercontent.com/49531960/76429434-697fba00-6385-11ea-870e-139955481487.png)

@jagraj

**Comment by jgwest** [comment]
It wouldn't surprise me if this is Windows only, in that Windows doesn't allow you to delete open files (unlike other platforms).

**Comment by mattcolegate** [comment]
Testing with latest Codewind and Codewind-vscode.
1. Brought up Codewind and VS Code
1. Created a Default Microclimate project
1. Waited for it to come up
1. Opened the performance dashboard
1. Ran a default load test to completion; data displayed on Performance dashboard
1. Right clicked project in IDE and selected 'Remove Project'

Project did fail to remove, but failed with a better error message and did not get rid of the project.
![image](https://user-images.githubusercontent.com/15730986/79346145-ea544780-7f29-11ea-8110-e40e923434d5.png)

Is the expected behaviour to be able to remove a project cleanly once load testing is complete?

**Comment by mattcolegate** [comment]
Ah I see - it's removed the underlying files and containers so although the project looks ok, it is in fact in an illegal state and now can't be removed further

**Comment by mattcolegate** [comment]
I've played around with the scenario a bit, and I believe it's something to do with the IDE - I got into the error state, closed VS Code and then re-opened it (leaving all other containers up and running, with the performance dashboard displayed). The deleted project no longer appears in the Codewind list, and I am able to delete the underlying project from the file system. The only file that was left in the project directory was the .project file - could the IDE be holding onto this for some reason? Perhaps waiting for a message from the performance dashboard that only gets issued when the dashboard is closed?
@sujeilyfonseca can you recreate and confirm that shutting down the IDE removes the lock on the file system?
@tetchel can you or someone from vsCode IDE land take a look please?

**Comment by tetchel** [comment]
What's the .project file? The projectID.json file? The IDE definitely does not use that directly; it only reads it through PFE.

**Comment by mattcolegate** [comment]
Hmm, in that case I wonder if PFE is holding it open for the IDE

**Comment by sujeilyfonseca** [comment]
@mattcolegate I confirm I was able to reproduce the same scenario you described. After restarting VS Code, the deleted project no longer appeared, and I was able to remove the project directory from the system. 

The message also looked better from me:

![Screen Shot 2020-04-15 at 11 51 54 AM](https://user-images.githubusercontent.com/49531960/79359208-4ad68b00-7f10-11ea-8a4a-e8bd351b27e0.png)

As an additional note, if you close the Performance Dashboard before attempting to delete the project, you get the same error, but there is no .project file left in the system directory after restarting the IDE. There is only an empty directory:

![Screen Shot 2020-04-15 at 11 53 49 AM](https://user-images.githubusercontent.com/49531960/79359579-c8020000-7f10-11ea-90c3-c3a00c40962d.png)

![Screen Shot 2020-04-15 at 11 53 56 AM](https://user-images.githubusercontent.com/49531960/79359589-c9cbc380-7f10-11ea-8831-882c42e21111.png)

**Comment by mattcolegate** [comment]
Still investigating this - PFE is holding the project file open but I've not narrowed down why yet. However, in an interesting development, I tried recreating this using Eclipse as the IDE and was unable to - the project was deleted cleanly. This makes me further suspect that VSCode plugin has something to do with this issue. @sujeilyfonseca, if you've got the time, I would be grateful if you could test the scenario with Eclipse IDE to confirm.

**Comment by tobespc** [comment]
PFE doesn't access the local disk anymore, so is the file that is being kept open in the pfe container or in the users directory ?

**Comment by sujeilyfonseca** [comment]
@mattcolegate, I tried to recreate the scenario using the Eclipse IDE, but I was unable to reproduce the problem.

**Comment by mattcolegate** [comment]
Yesterday I attempted to recreate this in front of @tobespc and @stalleyj but I wasn't able to - and I think it's down to the fact that I was using a VSCode session that already had files loaded into the workspace. I've now managed to recreate starting with a VSCode that had an empty workspace - when the project was created it automatically was added to the workspace. I used a program called Process Explorer to list the processes holding the project folder open on disk and they are all VSCode processes:
![image](https://user-images.githubusercontent.com/15730986/79979411-5f2f0080-8499-11ea-9d34-8a56d9c8da27.png)
![image](https://user-images.githubusercontent.com/15730986/79979689-d2d10d80-8499-11ea-89f1-de0feec964ac.png)
![image](https://user-images.githubusercontent.com/15730986/79979777-f300cc80-8499-11ea-8449-5f2995892c54.png)
I don't know why I saw a PFE process before using the folder and not now - I can confirm that I can shutdown and restart Codewind containers without freeing up the folder, so I'm confident now that it is VSCode holding the folder open. @tetchel, this defect should probably be tackled by you or your team.
Recreation steps:
1. Start a clean VSCode session - no workspace loaded
1. Start Codewind
1. Create a default microprofile project and wait until it is running - note that the project folder is added to the VSCode workspace
1. Open the Performance dashboard and run a default load test - wait for it to complete and data to be shown on the dashboard
1. Go back to VSCode and remove the project (including the directory on the file system). Error occurs deleting the project folder on the file system.

**Comment by tobespc** [comment]
@tetchel please can you take a look

**Comment by tetchel** [comment]
I see the file being closed correctly on my (mac) machine. I'll try windows later. The filesystem API I'm using is supposed to close the stream automatically so I'm not sure what may be causing this.

**Comment by tetchel** [comment]
OK, so the problem here seems to be that the VS Code java extension is keeping a reference to some files which prevents them from being deleted (on windows only). 

The solution is to remove the project folder from the workspace before deleting the files (currently the project folder is removed after deletion). But, if the project is the last workspace folder, removing that folder will cause the workspace to reload which will result in the extension 'forgetting' it has to delete the project. 

Trying to figure out how to handle this...

**Comment by sujeilyfonseca** [comment]
Thank you for working on this issue, @mattcolegate, @tobespc, and @tetchel!

I verified this issue with the latest Codewind 0.12.0, and I can no longer reproduce the problem.
